classes/plain.lua

```diff
@@ -260,20 +248,111 @@ function class:registerCommands ()
     SILE.call("font", { language = "und" }, content)
   end)
 
+  self:registerCommand("center", function (_, content)
+    if #SILE.typesetter.state.nodes ~= 0 then
+      SU.warn("\\center environment started after other nodes in a paragraph, may not center as expected")
+    end
+    SILE.settings:temporarily(function ()
+      local lskip = SILE.settings:get("document.lskip") or SILE.nodefactory.glue()
+      local rskip = SILE.settings:get("document.rskip") or SILE.nodefactory.glue()
+      SILE.settings:set("document.parindent", SILE.nodefactory.glue())
+      SILE.settings:set("current.parindent", SILE.nodefactory.glue())
+      SILE.settings:set("document.lskip", SILE.nodefactory.hfillglue(lskip.width.length))
+      SILE.settings:set("document.rskip", SILE.nodefactory.hfillglue(rskip.width.length))
+      SILE.settings:set("typesetter.parfillskip", SILE.nodefactory.glue())
+      SILE.settings:set("document.spaceskip", SILE.length("1spc", 0, 0))
+      SILE.process(content)
+      SILE.call("par")
+    end)
+  end, "Typeset its contents in a centered block (keeping margins).")
+
   self:registerCommand("raggedright", function (_, content)
-    SILE.call("ragged", { right = true }, content)
-  end)
+    SILE.settings:temporarily(function ()
+      local lskip = SILE.settings:get("document.lskip") or SILE.nodefactory.glue()
+      local rskip = SILE.settings:get("document.rskip") or SILE.nodefactory.glue()
+      SILE.settings:set("document.lskip", SILE.nodefactory.glue(lskip.width.length))
```

Comment:
So the question here (and on raggedleft) likewise is whether we should cancel parindent.
The old behavior was to cancel it, now we do not -- and it's a breaking change.

None of these behaviors is fully satisfying...
In ragged environments, one still needs to be able to see where paragraphs start... There are _usually_ two possibilities here:
- Either keeping a paragraph indent
- Or using a non-null paragraph skips

As I see it, there could be two approaches here:
- Either we let the user independently cancel the parindent if they do not want it.
- Or we check if there's a (fixed part) parskip and take decision upon it. But it might be too clever?
  ```lua
      -- If paragraphs are separated by a skip, the usual habit is to cancel the
      -- paragraph indentation.
      local pskip = SILE.settings:get("document.parskip") or SILE.nodefactory.glue()
      if pskip.height:tonumber() > 0 then
        SILE.settings:set("document.parindent", SILE.nodefactory.glue())
        SILE.settings:set("current.parindent", SILE.nodefactory.glue())
      end
   ```

Reply:
I don't like the "too clever" approach, given how subjective the right answer is it will be wrong for same use cases.

I also don't like automatically forcing one style on folks just because they switched an unrelated setting.

My suggestion would be to go ahead with the breaking change and start not touching the parindent setting just because ragged was turned on. We can throw a warning about how to enable it manually for folks whose documents may re-flow.

Reply:
I reviewed all the tests and found two types of changes: fixes to bugs we didn't even know existed because the previous resetting behavior was making them and expected changes based on new behavior. I updated some of them using `packages.retrograde` to keep the original test purpose intact while others that were incidental I just updated the new expectations (after verifying that the retrograde package restored the original expectations).